### PR TITLE
fix(retrieve): recover graph-coherent evidence skeletons

### LINKED ISSUE #625: [P0][Retrieval]: Replace phrase-gated full-flow recovery with generic evidence-skeleton selection
## Summary

`@lubab/madar@0.40.0-beta.3` passes the frozen #622 fixture but still fails a semantically equivalent natural prompt on the released 12,344-node / 32,788-edge JavaScript/TypeScript graph. This is an architectural retrieval regression, not missing graph data or insufficient budget.

Keep #622 and #623 as immutable historical receipts. This issue owns the systemic correction; do not open prompt-specific child bugs.

## Released field evidence

User prompt:

```text
explain how idea report generating is working
```

Claude call 1:

```text
How does idea report generation work end to end — pipeline stages, assembly, and persistence?
```

Observed result:

- 2 files / 2 snippets / 703 serialized tokens
- `PipelineAssemblyModule`
- `hasPersistedAnalysisData`
- one disconnected boundary between those unrelated endpoints
- zero relationships

Claude call 2:

```text
What is the pipeline orchestrator that runs idea analysis stages and enqueues jobs after an idea is created?
```

Observed result:

- 11 files / 14 snippets / 3,996 serialized tokens / truncated
- some planner/orchestrator evidence
- unrelated Runway, UI-stage, retry, recovery, and index-management symbols
- zero relationships
- missing the request/start, section-research, assembly, and DB-sync spine

Claude then broadly listed the pipeline tree and read roughly 3,600 source lines. The final answer was accurate because of filesystem fallback, not because Madar supplied a usable evidence skeleton.

Control: #622's exact frozen long query on the same graph returns the correct ten-stage spine in 3,669 tokens with five authenticated relationships. Therefore the graph, index semantics, and 4,000-token ceiling are not the blocker.

## Confirmed causes

1. Equivalent phrasing and punctuation select different algorithms in `rank.ts`.
2. Comma parsing can discard the lead clause as a local obligation and select isolated trailing-clause owners.
3. The #622 stage recovery is coupled to literal `flow` / `from ... to|through` / `between` grammar and one worker-registration/cycle topology.
4. Ranking rewards all-edge degree while traversal can return only authenticated `calls` / `enqueues_job` causal edges.
5. Generic lexical diversity outranks graph coherence and captures unrelated components.
6. `slice.ts` packs priority nodes before relationships, allowing snippets to consume the full budget and silently evict every edge.
7. Any nonempty node set is reported as `evidence` even when it cannot answer the requested flow.

## Required architecture

Use one deterministic retrieval pipeline for all questions:

```text
concept groups -> bounded candidate subgraph -> graph-coherent evidence skeleton/forest -> atomic evidence-bundle packing -> structural coverage check
```

- Punctuation creates soft concept groups but never activates another retrieval engine.
- Generate multiple lexical candidates per concept and expand class/module hits to executable owners.
- Separate navigation-only topology from causal output evidence. Navigation facts may locate owners but must never be emitted as invented causal transitions.
- Select bundles by concept coverage, authenticated path coverage, executable-owner value, disconnected-component penalty, redundancy, and token cost.
- Use bounded directional reachability, not only direct selected-node adjacency.
- Preserve honest asynchronous gaps as exact disconnected producer/consumer boundaries.
- Pack atomic facts in order: causal edge plus endpoints; disconnected pair plus boundary; required intermediates; supplemental anchors.
- Permit at most one internal bounded substitution/closure when concepts are uncovered or multiple selected anchors yield zero edges.

### Real released incident

- Freeze the two exact failing calls above and the known passing #622 control before implementation.
- The end-to-end query returns evidence for request/start, orchestration/planning, research, assembly/final report, and persistence.
- The orchestrator follow-up includes start/enqueue, orchestration, research, and assembly anchors with non-empty authenticated relationships.
- All available causal relationships survive slicing.
- Unauthenticated async handoffs are exact disconnected boundaries; no edge is fabricated.
- No UI persistence helper, Runway controller, retry parser, or unrelated recovery helper.
- No broad filesystem restart is required by the packed-candidate Claude/Codex canary.

### Generalization

- Distant paraphrases, punctuation changes, and clause reordering retain every required stage category.
- Alpha-renaming symbols/files/framework suffixes does not change graph-grounded phase coverage.
- Deterministic 10k+ node distractor overlays do not remove a critical phase or introduce more than two unrelated files.
- Synchronous chains, queued handoffs, fan-out/fan-in, cycles, and disconnected handoffs remain honest.
- Focused locator questions remain focused and do not expand into the full skeleton.

### Hard limits and quality gates

- Existing ceilings remain unchanged: at most 12 files, 25 snippets, 4,000 serialized tokens, and one closure pass.
- Worst-paraphrase required-phase recall is 100% for blocking cases.
- Handoff truth and evidence authenticity are 100%; zero invented or reversed edges.
- Selected-file precision is at least 70% and false-ready count is zero.
- Loaded-graph p95 remains under the existing 500 ms reference gate with no material memory regression.
- Full tests, typecheck, build, checkout/direct/CLI/MCP parity, package isolation, independent exact-head review, and all protected-`next` CI gates pass.

## Implementation boundary

Expected production surfaces:

- `src/domain/query/rank.ts`
- `src/domain/query/traverse.ts`
- `src/domain/query/slice.ts`
- internal types in `src/domain/query/types.ts`
- orchestration in `src/application/retrieve-context.ts`
- focused tests and development-only evaluation tooling

## Non-goals / stop conditions

- No query-by-query synonym or regex patch.
- No repository names, paths, expected symbols, or prompt literals in production code.
- No new dependency, embedding/model call, second retrieval engine, recursive planner, provider traffic, or paid evaluation.
- No budget, file, snippet, closure, timeout, package, or provider ceiling increase.
- No graph/index semantic change unless a controlled experiment proves required evidence is absent. Current control evidence says it is present.
- No npm publication, GitHub Release, MCP Registry publication, tag, stable/`latest`, or `main` action under this implementation issue.
- If the implementation requires a public schema or graph/index semantic change, stop and request a separately reviewed architecture amendment.

## Delivery

- Base all work on the latest protected `next` commit.
- Use an isolated worktree and preserve user-owned root changes.
- One architecture implementation PR against `next` after exact-snapshot independent review.
- A separate prerelease PR/canary may be proposed only after this issue's gates pass.

Related: #618, #622, #623, #624.

### PR DESCRIPTION
**Pull request by mohanagy**
Closes #625.

## Summary

- replace phrase-gated full-flow recovery with one deterministic concept-group -> candidate-subgraph -> evidence-skeleton/forest -> exact bundle-pack -> structural-coverage pipeline
- preserve authenticated causal edges and honest disconnected handoffs across sequential, fan-out/fan-in, cyclic, scoped, test-domain, and 2,000-parallel-edge cases
- keep exact locators focused and prevent production queries from authenticating through test-only topology
- use exact canonical-JSON splice token deltas so oversized relationships cannot block later fitting evidence
- record the first-head CI/review stop and the exact corrective evidence without widening any constraint

## Exact corrective candidate

- protected `next` base: `ece7d0d02643ecec08bd91aa904a4514aa845f42`
- frozen source/test snapshot E: `4285bdf3f335a31b9c5176cb4215db3aabc5e787`
- E tree / `src` tree / retrieval-test tree: `d6c04a8189e9db73a709b259b64d07dc8d09f7b7` / `3d146a24b2fd457159493be2e68c4c7c57c550f5` / `0942cb5258a022fd5a929d3ba32addd42e3998fd`
- final PR head F / tree: `3d194b62aba3324e93bcf01c77cf508edea70a6f` / `15565d356d947b2b195e7e0ec4871094b60b8493`
- E-to-F governance-only diff: exactly the removal manifest, scorecard, roadmap, RFC design, and governance test; SHA-256 `79046909f3e297ef782ff26967735def4697bf322bf335e995ab65847913155b`
- production scope: exactly five existing paths, 43 production TypeScript files / 12,454 LOC, `+1,409/-1,102/net +307`; binary diff SHA-256 `b16c0e04fccc044d421cb594fe90504a56e5197f2702e826d4a98394d6450131`
- package: 102 entries / 159,980 packed / 639,930 unpacked bytes; shasum `398f08c8315a3b2ba8fb1c0ec22e17d8bf2c3c05`; tarball SHA-256 `d30d7990fe217bd8b50a8fb2d0085b1864ce4a8a1725731424bde00957934b76`

## Validation

- real GoValidate graph: all five field-incident variants pass 108/108 exact invocations and return the ten-stage spine, 5 authenticated calls, 4 honest disconnected boundaries, 10 files / 10 snippets / 3,669 tokens / 1 closure / not truncated
- accepted warmed reference gate: first 20 p50 139.214 ms / p95 161.626 / max 162.958; 100-call extension p50 129.501 / p95 145.671 / max 162.958; hosted CI makes no reference-hardware latency claim
- corrected three-trial memory gate: after two macrotask yields and forced GC, protected-base median heap is 115,398,016 bytes and candidate median is 114,099,800 bytes (-1,298,216); the 31,549,504-byte same-job corpus is reclaimable WeakRef keep-during-job behavior, not a retained leak
- post-collection cold candidate median 660.295 ms is recorded honestly as diagnostic outside the inherited warmed-reference protocol
- exact candidate gate 128/128; full coverage suite 76 files / 684 tests at 83.74% statements / 75.41% branches / 91.00% functions / 87.56% lines
- typecheck, production/evaluation builds, release hygiene, Registry validation, packed parity, npm audit with zero vulnerabilities, and independent source/test plus final-governance reviews pass

## First-head stop and remaining gates

First head `eb664013a399bd4ef970915ce9382f841395ebfa` was not merged: all six hosted jobs failed invalid cold single-sample timing assertions, and CodeRabbit opened five actionable threads. The correction removes only those invalid hosted timing claims, fixes sentence-boundary casing and locator formatting, and adds exact hard-budget regressions. It changes no accepted performance ceiling, retrieval timeout, budget, dependency, graph/index/schema behavior, package ceiling, or platform rule.

All six CI jobs, fresh CodeRabbit disposition, independent final-head review, and zero unresolved threads remain blocking on exact head F.

No npm/product publication, GitHub Release, Registry metadata, tag, stable/`latest`, or `main` action is authorized or performed.

## Summary by CodeRabbit

* **New Features**
  * Added structural-coverage controls to retrieval evidence slicing to better handle incomplete structure requirements.
* **Bug Fixes**
  * Improved anchor prioritization and ranking for paraphrased and complex queries.
  * More reliable traversal and evidence boundary handling, including disconnected details and structural “missing” reporting.
* **Performance**
  * Enhanced budget-aware evidence selection with improved token/accounting stability.
* **Release**
  * Published version 0.40.0-beta.3 via the next release channel.

### CODE REVIEW

**Comment by mohanagy** [comment]
## Independent exact-head review receipt

Reviewed exact PR head `eb664013a399bd4ef970915ce9382f841395ebfa` / tree `ebc3668ad0f6a5da09f56208909b92d6641a3b73` against protected `next` base `ece7d0d02643ecec08bd91aa904a4514aa845f42`.

Verdict: **no P0/P1 blocker**.

The review independently verified that only the five governance files differ after frozen source/test snapshot C; every frozen source, fixture, and functional-test path remains byte-identical to C. It reproduced the exact five-path production delta and SHA-256, the 102 / 159,961 / 639,876 package and all hashes, fixture-derived 15/14/1/5 query counts, synthetic 6/3 versus live GoValidate 5/4 topology separation, performance/memory receipts, and the 2,000-parallel-edge exact-pack behavior. The exact-head five-file gate passes 124/124 and the governance gate passes 18/18.

This review does not replace or pre-approve the remaining gates: all six CI jobs, CodeRabbit’s actual disposition, zero unresolved review threads, and unchanged-head squash merge are still required.

**Comment by mohanagy** [comment]
## Corrective exact-head receipt — retrieval regression #625

The first PR head remains an immutable stop. This receipt records the corrective candidate now published to PR #626; it does not claim CI, CodeRabbit, zero-thread, merge, or publication completion.

- Protected base / target: `next` at `ece7d0d02643ecec08bd91aa904a4514aa845f42` / `next`.
- First stopped PR head: `eb664013a399bd4ef970915ce9382f841395ebfa`; all six jobs in run 30527791296 failed invalid cold single-sample timing assertions, CodeRabbit opened five actionable threads, and no merge was attempted.
- Frozen corrective source/test snapshot E: `4285bdf3f335a31b9c5176cb4215db3aabc5e787`; tree `d6c04a8189e9db73a709b259b64d07dc8d09f7b7`; `src` tree `3d146a24b2fd457159493be2e68c4c7c57c550f5`; retrieval-test tree `0942cb5258a022fd5a929d3ba32addd42e3998fd`.
- Exact corrective PR head F: `3d194b62aba3324e93bcf01c77cf508edea70a6f`; tree `15565d356d947b2b195e7e0ec4871094b60b8493`.
- E-to-F governance diff: exactly five governance paths; SHA-256 `79046909f3e297ef782ff26967735def4697bf322bf335e995ab65847913155b`.
- Production: exactly five existing paths, no added production file, 43 TypeScript files / 12,454 LOC, `+1,409/-1,102/net +307`; binary diff SHA-256 `b16c0e04fccc044d421cb594fe90504a56e5197f2702e826d4a98394d6450131`.
- Package: 102 entries / 159,980 packed / 639,930 unpacked bytes under unchanged 102 / 165,000 / 640,000 ceilings; shasum `398f08c8315a3b2ba8fb1c0ec22e17d8bf2c3c05`; integrity `sha512-X1DWZZdJmq3Y2nZzfPEQpWkU4c/eLSguPCxK4F6YSCx/iCqOHK0zrA4jHpPVhYNufDrpcIeyNeW5EahZ+Q7kPQ==`; tarball SHA-256 `d30d7990fe217bd8b50a8fb2d0085b1864ce4a8a1725731424bde00957934b76`.
- Real graph: five incident phrasings pass 108/108 exact invocations with 10 stages/files, 10 snippets, 3,669 tokens, five authenticated calls, four disconnected handoffs, one closure, and no truncation.
- Warm reference: first 20 p50 139.214 ms / p95 161.626 / max 162.958; 100-call extension p50 129.501 / p95 145.671 / max 162.958, below the unchanged strict `<500 ms` gate. Hosted CI makes no reference-hardware latency claim.
- Corrected memory protocol: three separate processes per head, three five-query warmups, five forced-GC rounds, and two macrotask yields. Median post-yield heap is 115,398,016 bytes on the protected base and 114,099,800 on the candidate (-1,298,216). The 31,549,504-byte same-job corpus is reclaimable WeakRef keep-during-job behavior, not a retained leak. The 660.295 ms post-collection cold candidate median is diagnostic outside the inherited warmed-reference gate.
- Local gates: exact candidate 128/128; full coverage suite 76 files / 684 tests at 83.74% statements / 75.41% branches / 91.00% functions / 87.56% lines; packed parity, typecheck, both builds, release hygiene, Registry validation, npm audit zero, source/test review, and final-governance review pass.

Still blocking on exact head F: all six protected CI jobs, fresh CodeRabbit disposition, independent final-head review, and zero unresolved review threads. Merge remains forbidden until all pass.

No npm/product publication, GitHub Release, Registry metadata, tag, stable/`latest`, or `main` action is authorized or performed.

File: src/domain/query/slice.ts [hunk 1]
```diff
@@ -12,275 +14,381 @@ export interface SliceEvidenceInput {
...
+function addEdgeTokens(current: number, delta: number): number {
+  const body = current - countTokens(String(current)) + delta
+  let tokens = body
+  for (let pass = 0; pass < 16; pass += 1) {
+    const observed = body + countTokens(String(tokens))
+    if (observed === tokens) return tokens
+    tokens = observed
+  }
+  throw new Error('Unable to stabilize retrieve serialized token count')
 }
```
[the remark this answers is not in the document]
Comment (mohanagy) [reply src/domain/query/slice.ts:83]: Verified against corrective head 3d194b62aba3324e93bcf01c77cf508edea70a6f: the proposed over-budget path is not reproducible. Incremental token accounting is confined to stable edge-only insertions; committed punctuation, escaping, Unicode, exact-boundary, maximum-question/minimum-budget tests, 50,000 independent splice comparisons, and 120 randomized full-slice trials found zero token or budget violations. A speculative removal clamp was rejected because it broke atomic packing and the hard minimum-budget envelope. No budget or ceiling is widened.

File: tests/unit/retrieve-evidence-skeleton-adversarial.test.ts [hunk 2]
```diff
@@ -0,0 +1,1100 @@
...
+      budget: 4_000,
+    })
+    const elapsed = performance.now() - started
+
+    expect(ranked.priorityAnchorIds).toEqual([
+      'chain-00000',
+      'chain-12343',
+    ])
+    expect(elapsed).toBeLessThan(500)
+  })
```
[the remark this answers is not in the document]
Comment (mohanagy) [reply tests/unit/retrieve-evidence-skeleton-adversarial.test.ts]: Corrected on exact head 3d194b62aba3324e93bcf01c77cf508edea70a6f. The cold single-sample hosted assertion was removed because it did not implement the inherited warmed loaded-graph p95 protocol, while the exact 12,344-node endpoint assertions remain. The independent reference gate passes at 161.626 ms p95 against the unchanged strict <500 ms limit; hosted CI now claims correctness and portability only.

**Review by mohanagy — COMMENTED** [review]
Independent exact-head review — NO P0/P1 blocker on 3d194b62aba3324e93bcf01c77cf508edea70a6f / tree 15565d356d947b2b195e7e0ec4871094b60b8493. The review reproduced the frozen source/test identities, 43-file/12,454-LOC inventory, +1,409/-1,102/net+307 diff, package 102/159,980/639,930 and hashes, real-graph/synthetic separation, warmed <500 ms reference evidence, corrected two-yield WeakRef memory arithmetic, first-head six-job stop, five-thread disposition, unchanged constraints, and clean governance. Governance tests pass 18/18 and git diff --check is clean. CI and fresh CodeRabbit remain separate blocking gates.

**Comment by mohanagy** [comment]
## Completion receipt — retrieval regression #625

PR #626 was squash-merged into protected `next`.

- Reviewed head: `3d194b62aba3324e93bcf01c77cf508edea70a6f`
- Reviewed head tree: `15565d356d947b2b195e7e0ec4871094b60b8493`
- Merge commit: `b6562b715133304bd46e537b6f39008bc1e02095`
- Merge tree: `15565d356d947b2b195e7e0ec4871094b60b8493` — exactly equal to the reviewed head tree
- Sole parent: protected `next` base `ece7d0d02643ecec08bd91aa904a4514aa845f42`
- Exact-head CI: all six required jobs passed in run 30533140531
- CodeRabbit: completed successfully on the exact final head; no actionable comment remained, two non-blocking nitpicks were recorded, and this was not a non-default-base skip
- Independent exact-head review: no P0/P1 blocker
- Review threads at merge: zero unresolved
- Package: 102 entries / 159,980 packed / 639,930 unpacked bytes under unchanged 102 / 165,000 / 640,000 ceilings
- Real graph: five incident phrasings passed 108/108 exact invocations with the exact ten-stage spine, five authenticated calls, four honest disconnected handoffs, 10 files / 10 snippets / 3,669 tokens / one closure / not truncated
- Warm reference p95: 161.626 ms against unchanged strict `<500 ms`
- Corrected post-yield forced-GC memory comparison: candidate median heap 1,298,216 bytes below protected base; no material retained-heap regression

The systemic evidence-skeleton pipeline is complete. The issue is closed as completed.

No npm/product publication, GitHub Release, Registry metadata, tag, stable/`latest`, or `main` action was performed or authorized by this merge.
